File: ios/AlphaCityStampTour/AlphaCityStampTour/Network/DTO/HomeDTO.swift
//
//  HomeDTO.swift
//  AlphaCityStampTour
//

import Foundation

// MARK: - 배너

struct BannerData: Decodable, Identifiable {
    let id: Int
    let title: String
    let imageUrl: String
}

// MARK: - 프로그램

struct ProgramData: Decodable, Identifiable {
    let id: Int
    let name: String
    let description: String?
    let category: String?
    let subcategory: String?
    let hasCoupon: Bool?
    let imageUrl: String?
    let operatingHours: String?
    let location: String?
    let latitude: Double?
    let longitude: Double?
    let phone: String?
    let speaker: String?
    let ownerName: String?
    let storeCode: String?
    let operatingDays: String?
    let startDate: String
    let endDate: String
    let status: String
    let events: [EventData]?
    let storeCoupons: [StoreCouponInfo]?
    let stores: [ProgramStoreData]?
}

struct StoreCouponInfo: Decodable {
    let storeId: Int
    let storeName: String
    let couponId: Int
    let couponName: String
    let couponDescription: String?
}

struct ProgramStoreData: Decodable, Identifiable {
    let id: Int
    let name: String
}

// MARK: - 미션

struct MissionData: Decodable, Identifiable {
    let id: Int
    let name: String
    let type: String
    let placeId: Int?
    let programId: Int?
    let stampId: Int?
    let question: String?
    let answer: String?
    let options: [String]?
    let stayMinutes: Int?
    let place: MissionPlace?
    let isCompleted: Bool?

    struct MissionPlace: Decodable {
        let name: String
        let latitude: Double?
        let longitude: Double?
    }
}

// MARK: - 미션 완료 결과

struct MissionCompletionData: Decodable {
    let id: Int
    let missionId: Int
    let userId: Int
    let completedAt: String?
    let stamp: StampData?
}

// MARK: - 이벤트

struct EventData: Decodable, Identifiable {
    let id: Int
    let name: String
    let description: String?
    let imageUrl: String?
    let type: String
    let startDate: String
    let endDate: String
    let reward: String?
    let winnerCount: Int?
    let participantLimit: Int
    let participantCount: Int?
    let status: String
    let program: EventProgram?
    let isParticipated: Bool?
    let myRaffleNumber: Int?
    let price: Int?
    let duration: Int?
    let capacity: Int?
    let location: String?

    struct EventProgram: Decodable {
        let name: String
    }
}

// MARK: - 이벤트 참여 결과

struct EventParticipationData: Decodable {
    let id: Int
    let eventId: Int
    let userId: Int
    let raffleNumber: Int?
    let joinedAt: String
}

// MARK: - 스탬프

struct StampData: Decodable, Identifiable {
    let id: Int
    let name: String
    let conditionType: String
    let conditionDetail: String?
    let imageUrl: String?
}

// MARK: - 유저 스탬프 수집 기록

struct UserStampData: Decodable, Identifiable {
    let id: Int
    let stampId: Int
    let collectedAt: String?
}

// MARK: - 쿠폰

struct CouponListResponse: Decodable {
    let coupons: [CouponData]
    let redeemedCouponIds: [Int]
    let availableStamps: Int?
}

struct CouponData: Decodable, Identifiable {
    let id: Int
    let name: String
    let description: String?
    let requiredStamps: Int
    let validUntil: String
    let imageUrl: String?
    let programId: Int?
}

struct UserCouponData: Decodable, Identifiable {
    let id: Int
    let userId: Int
    let couponId: Int
    let code: String
    let status: String
    let storeId: Int?
    let requestedAt: String?
    let usedAt: String?
    let createdAt: String?
}

struct MyCouponData: Decodable, Identifiable {
    let id: Int
    let couponId: Int
    let name: String
    let description: String?
    let imageUrl: String?
    let code: String
    let status: String
    let validUntil: String
    let storeName: String?
    let requestedAt: String?
    let usedAt: String?
    let createdAt: String?
}

// MARK: - 활동 이력

struct ActivityItemDTO: Decodable {
    let type: String
    let subType: String?
    let title: String
    let date: String
    let validUntil: String?
}

// MARK: - 유저 프로필

struct UserProfileData: Decodable {
    let id: Int
    let email: String
    let nickname: String
    let name: String?
    let phone: String?
    let address: String?
    let addressDetail: String?
    let profileImage: String?
    let provider: String?
    let birthDate: String?
    let gender: String?
    let stampCount: Int?
    let couponCount: Int?
    let missionCount: Int?
}

File: compose-web-test/src/commonMain/kotlin/network/dto/HomeDto.kt
package com.alphacity.stamptour.network.dto
import kotlinx.serialization.Serializable

@Serializable
data class BannerItem(
    val id: Int,
    val title: String,
    val imageUrl: String,
    val linkType: String? = null,
    val linkId: Int? = null,
)
@Serializable
data class FestivalItem(
    val id: Int,
    val name: String,
    val description: String? = null,
    val imageUrl: String? = null,
    val bannerUrl: String? = null,
    val startDate: String,
    val endDate: String,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val address: String? = null,
    val sortOrder: Int = 0,
    val isActive: Boolean = true,
    val status: String,
)
@Serializable
data class FestivalDetail(
    val id: Int,
    val name: String,
    val description: String? = null,
    val imageUrl: String? = null,
    val bannerUrl: String? = null,
    val startDate: String,
    val endDate: String,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val address: String? = null,
    val sortOrder: Int = 0,
    val isActive: Boolean = true,
    val status: String,
    val programs: List<FestivalProgramSummary> = emptyList(),
)
@Serializable
data class FestivalProgramSummary(
    val id: Int,
    val name: String,
    val category: String? = null,
    val imageUrl: String? = null,
    val startDate: String,
    val endDate: String,
    val status: String? = null,
    val location: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
)
@Serializable
data class ProgramItem(
    val id: Int,
    val festivalId: Int? = null,
    val name: String,
    val description: String? = null,
    val category: String? = null,
    val subcategory: String? = null,
    val hasCoupon: Boolean? = false,
    val ttsUrl: String? = null,
    val imageUrl: String? = null,
    val operatingHours: String? = null,
    val location: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val speaker: String? = null,
    val startDate: String,
    val endDate: String,
    val status: String,
    val phone: String? = null,
    val ownerName: String? = null,
    val storeCode: String? = null,
    val operatingDays: String? = null,
    val events: List<EventItem>? = null,
    val storeCoupons: List<StoreCouponInfo>? = null,
    val stores: List<ProgramStoreItem>? = null,
)
@Serializable
data class StoreCouponInfo(
    val storeId: Int,
    val storeName: String,
    val couponId: Int,
    val couponName: String,
    val couponDescription: String? = null,
)
@Serializable
data class ProgramStoreItem(
    val id: Int,
    val name: String,
)
@Serializable
data class EventItem(
    val id: Int,
    val name: String,
    val description: String? = null,
    val imageUrl: String? = null,
    val type: String,
    val startDate: String,
    val endDate: String,
    val reward: String? = null,
    val winnerCount: Int? = null,
    val participantLimit: Int = 0,
    val participantCount: Int = 0,
    val status: String,
    val program: EventProgram? = null,
    val isParticipated: Boolean? = null,
    val myRaffleNumber: Int? = null,
    val price: Int? = null,
    val duration: Int? = null,
    val capacity: Int? = null,
    val location: String? = null,
)
@Serializable
data class EventParticipationResult(
    val id: Int,
    val eventId: Int,
    val userId: Int,
    val raffleNumber: Int? = null,
    val joinedAt: String,
)
@Serializable
data class EventProgram(
    val name: String,
)
@Serializable
data class StampItem(
    val id: Int,
    val festivalId: Int? = null,
    val programId: Int? = null,
    val name: String,
    val conditionType: String,
    val conditionDetail: String? = null,
    val imageUrl: String? = null,
)
@Serializable
data class UserStampItem(
    val id: Int,
    val stampId: Int,
    val collectedAt: String? = null,
    val stamp: StampItem? = null,
)
@Serializable
data class MissionItem(
    val id: Int,
    val festivalId: Int? = null,
    val name: String,
    val type: String,
    val placeId: Int? = null,
    val programId: Int? = null,
    val stampId: Int? = null,
    val question: String? = null,
    val answer: String? = null,
    val options: List<String>? = null,
    val stayMinutes: Int? = null,
    val place: MissionPlace? = null,
    val isCompleted: Boolean? = null,
)
@Serializable
data class MissionCompletionResult(
    val id: Int,
    val missionId: Int,
    val userId: Int,
    val completedAt: String? = null,
    val stamp: StampItem? = null,
)

@Serializable
data class MissionPlace(
    val name: String,
    val latitude: Double? = null,
    val longitude: Double? = null,
)

@Serializable
data class CouponListResponse(
    val coupons: List<CouponItem>,
    val redeemedCouponIds: List<Int> = emptyList(),
    val availableStamps: Int = 0,
)

@Serializable
data class CouponItem(
    val id: Int,
    val festivalId: Int? = null,
    val name: String,
    val description: String? = null,
    val requiredStamps: Int,
    val validUntil: String,
    val imageUrl: String? = null,
    val programId: Int? = null,
)

@Serializable
data class UserCouponItem(
    val id: Int,
    val userId: Int,
    val couponId: Int,
    val code: String,
    val status: String,
    val storeId: Int? = null,
    val requestedAt: String? = null,
    val usedAt: String? = null,
    val createdAt: String? = null,
)

@Serializable
data class MyCouponItem(
    val id: Int,
    val couponId: Int,
    val name: String,
    val description: String? = null,
    val imageUrl: String? = null,
    val code: String,
    val status: String,
    val validUntil: String,
    val storeName: String? = null,
    val requestedAt: String? = null,
    val usedAt: String? = null,
    val createdAt: String? = null,
)

@Serializable
data class UserProfile(
    val id: Int,
    val email: String,
    val nickname: String,
    val name: String? = null,
    val phone: String? = null,
    val address: String? = null,
    val addressDetail: String? = null,
    val profileImage: String? = null,
    val provider: String? = null,
    val birthDate: String? = null,
    val gender: String? = null,
    val stampCount: Int? = null,
    val couponCount: Int? = null,
    val missionCount: Int? = null,
)

@Serializable
data class UpdateProfileRequest(
    val nickname: String,
    val name: String? = null,
    val currentPassword: String? = null,
    val newPassword: String? = null,
    val phone: String? = null,
    val address: String? = null,
    val addressDetail: String? = null,
    val birthDate: String? = null,
    val gender: String? = null,
    val profileImage: String? = null,
)

@Serializable
data class ActivityItemDto(
    val type: String,
    val subType: String? = null,
    val title: String,
    val date: String,
    val validUntil: String? = null,
)